src/Adapter/Order/OrderProductUpdater.php

```diff
@@ -0,0 +1,519 @@
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function syncCartAndOrderCartRules(Cart $cart, Order $order, bool $delete = false): Cart
+    {
+        Context::getContext()->cart = $cart;
+        CartRule::autoAddToCart();
+        CartRule::autoRemoveFromCart();
+
+        $orderCartRulesData = $order->getCartRules();
+        $orderCartRulesIds = array_map(
+            function (array $orderCartRuleData) { return (int) $orderCartRuleData['id_cart_rule']; },
+            $orderCartRulesData
+        );
+
+        $cartRules = $cart->getCartRules();
+        $cartRulesIds = array_map(
+            function (array $cartRule) { return (int) $cartRule['id_cart_rule']; },
+            $cartRules
+        );
+
+        if (count($orderCartRulesIds) === count($cartRulesIds)) {
+            return $cart;
+        }
+
+        if (count($orderCartRulesIds) > count($cartRulesIds)) {
+            if ($delete) { // Product deletion, the extra order cart rules have to be deleted
+                foreach ($order->getCartRules() as $orderCartRule) {
+                    foreach (array_diff($orderCartRulesIds, $cartRulesIds) as $idCartRule) {
+                        if ($idCartRule == $orderCartRule['id_cart_rule']) {
+                            $orderCartRule = new OrderCartRule($orderCartRule['id_order_cart_rule']);
+                            $orderCartRule->delete();
+                        }
+                    }
+                }
+            } else { // desynchronization of cartRules, need to be fixed
+                sort($orderCartRulesIds);
+                sort($cartRulesIds);
+
+                foreach (array_diff($orderCartRulesIds, $cartRulesIds) as $cartRuleId) {
+                    $cart->addCartRule($cartRuleId);
+                }
+            }
+        } else {
+            throw new \LogicException('Cart has more cart rules than order. This should never happen.');
+        }
+
+        return $cart;
+    }
+
+    /**
+     * @param Cart $cart
+     * @param Order $order
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function recalculateCartRules(Cart $cart, Order $order): void
+    {
+        $computingPrecision = $this->getPrecisionFromCart($cart);
+
+        foreach ($order->getCartRules() as $orderCartRuleData) {
+            $orderCartRule = new OrderCartRule((int) $orderCartRuleData['id_order_cart_rule']);
+            $idCartRule = (int) $orderCartRule->id_cart_rule;
+            $cartRule = new CartRule($idCartRule);
+            $values = [
+                'tax_incl' => Tools::ps_round($cartRule->getContextualValue(true), $computingPrecision),
+                'tax_excl' => Tools::ps_round($cartRule->getContextualValue(false), $computingPrecision),
+            ];
+
+            if (
+                ($values['tax_incl'] !== Tools::ps_round($orderCartRule->value, $computingPrecision)) ||
```

Comment:
Not sure if this check is necessary, we could update the order all the time (just one additional request is acceptable if it ensure stability) Besides the `$values` could be removed with simpler code:
```php
$orderCartRule->value = Tools::ps_round($cartRule->getContextualValue(true), $computingPrecision);
$orderCartRule->value_tax_excl = Tools::ps_round($cartRule->getContextualValue(false), $computingPrecision);
```